numpy/core/tests/test_multiarray.py

```diff
@@ -4380,6 +4380,12 @@ def setUp(self):
         self.omat = np.array([Decimal(repr(r)) for r in self.rmat.flat])
         self.omat = self.omat.reshape(4, 5)
 
+    def test_python_type(self):
+        for x in (np.float16(1.), 1, 1., 1+0j):
```

Comment:
To have a test that triggers the bug fixed in this PR, you may want to add `[np.float16(1.)]` to this list (i.e., a float16 wrapped in a list). The others don't run into the `is_float16_result` and `not hasattr(a, "dtype")` case.